include/iam/publicnodeclient.hpp

```diff
@@ -0,0 +1,118 @@
+class ConnectionCallbackItf {
+public:
+    virtual ~ConnectionCallbackItf() = default;
+
+    virtual bool OnConnected() = 0;
+};
+
+/**
+ * @brief Public node client interface.
+ */
+class PublicNodeClientItf : public ConnectionCallbackItf {
+public:
+    /**
+     * @brief Initialize the client.
+     *
+     * @param certContext certificate context.
+     * @param provisioningMode provisioning mode.
+     * @return Error error code.
+     */
+    Error Init(const CryptoCertContext& certContext, bool provisioningMode);
+
+    /**
+     * @brief Run the client.
+     *
+     * @param url URL of the public node.
+     * @param reconnectInterval interval for reconnection.
+     */
+    void Run(const std::string& url, aos::common::utils::Duration reconnectInterval);
+
+    /**
+     * @brief Close the client.
+     */
+    void Close();
+
+private:
+    using StreamPtr = std::unique_ptr<
+        grpc::ClientReaderWriterInterface<iamanager::v5::IAMOutgoingMessages, iamanager::v5::IAMIncomingMessages>>;
+
+protected:
+    StreamPtr mStream;
+
+private:
+    using PublicNodeService        = iamanager::v5::IAMPublicNodesService;
+    using PublicNodeServiceStubPtr = std::unique_ptr<PublicNodeService::StubInterface>;
+    using HandlerFunc              = std::function<bool(const iamanager::v5::IAMIncomingMessages&)>;
+
+    virtual Error GetCert(const std::string& certType, aos::iam::certhandler::CertInfo& certInfo)    = 0;
+    virtual bool  ProcessStartProvisioning(const iamanager::v5::StartProvisioningRequest& request)   = 0;
+    virtual bool  ProcessFinishProvisioning(const iamanager::v5::FinishProvisioningRequest& request) = 0;
+    virtual bool  ProcessDeprovision(const iamanager::v5::DeprovisionRequest& request)               = 0;
+    virtual bool  ProcessPauseNode(const iamanager::v5::PauseNodeRequest& request)                   = 0;
+    virtual bool  ProcessResumeNode(const iamanager::v5::ResumeNodeRequest& request)                 = 0;
+    virtual bool  ProcessCreateKey(const iamanager::v5::CreateKeyRequest& request)                   = 0;
+    virtual bool  ProcessApplyCert(const iamanager::v5::ApplyCertRequest& request)                   = 0;
+    virtual bool  ProcessGetCertTypes(const iamanager::v5::GetCertTypesRequest& request)             = 0;
+
+    void ConnectionLoop(const std::string& url, utils::Duration reconnectInterval) noexcept;
+    void HandleIncomingMessages();
+    bool RegisterNode(const std::string& url);
+    void InitializeHandlers();
+
+    std::vector<std::shared_ptr<grpc::ChannelCredentials>>                                      mCredentialList;
+    std::unordered_map<iamanager::v5::IAMIncomingMessages::IAMIncomingMessageCase, HandlerFunc> mHandlers;
+
+    std::unique_ptr<grpc::ClientContext> mRegisterNodeCtx;
+    PublicNodeServiceStubPtr             mPublicNodeServiceStub;
+
+    std::thread             mConnectionThread;
+    std::condition_variable mShutdownCV;
+    std::atomic<bool>       mShutdown {};
```

Comment:
i disagree because if you look in implementation of function 
`bool PublicNodeClientItf::RegisterNode(const std::string& url)` so when we iterate over mCredentialList we block mutex so in this case when someone wanna to close, it will block until we iterate over whole mCredentialList, I agree there could be possible to reduce scope, but I think it much better to use atomic, and this operation is cheap